### PR TITLE
Add USB Ethernet support to GNRC Border Router example

### PR DESCRIPTION
**Pull request by trickeydan**
### Contribution description

Add USB ethernet (CDC-ECM) support as an uplink to the `grnc_border_router` example. This allows using the usb ethernet drivers as an alternative to SLIP or ethos, which compared to ethos reduces latency over the link.

### Testing procedure

Flash the example to a board with USB support, for example `samr21-xpro`.

`sudo make UPLINK=usb-cdc-ecm all flash`

You will then need to attach the usb port to your computer, which should automatically detect and bring up the interface with link-locals on most distributions. Bear in mind that this may be a different USB port to the one that you used to flash the board.

I have successfully plugged in my `samr21-xpro` with two micro usb cables, one into the device port, and one into the edbg port for testing. This allows access to the serial terminal over the EDBG serial, without having to use the USB CDC-ACM driver for serial.

`sudo make UPLINK=usb-cdc-ecm term`

### Issues/PRs references

N/A

### CODE REVIEW

File: examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# USB Modules
+USEMODULE += usbus
+USEMODULE += auto_init_usbus
+USEMODULE += usbus_cdc_ecm
+
+ifeq (0,$(USE_DHCPV6))
```
Comment (benpicco) [inline examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf:6]: Why shouldn't your `start_network.sh` script also run for DHCPv6? You even added a `--use-dhcpv6` option.

File: examples/gnrc_border_router/Makefile [hunk 2]
```diff
@@ -11,8 +11,8 @@ RIOTBASE ?= $(CURDIR)/../..
 UPLINK ?= ethos
 
 # Check if the selected Uplink is valid
-ifeq (,$(filter ethos slip wifi,$(UPLINK)))
-  $(error Supported uplinks are `ethos`, `slip` and `wifi`)
+ifeq (,$(filter ethos slip usb-cdc-ecm wifi,$(UPLINK)))
+  $(error Supported uplinks are `ethos`, `slip`, `usb-cdc-ecm` and `wifi`)
```
Comment (benpicco) [inline examples/gnrc_border_router/Makefile:15]: Maybe just `cdc-ecm`? That's easier to type ;)

File: dist/tools/usb-cdc-ecm/start_network.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+setup_interface
+
+if [ ${USE_DHCPV6} -eq 1 ]; then
+    DHCPD="$(readlink -f "${USB_CDC_ECM_DIR}/../dhcpv6-pd_ia/")/dhcpv6-pd_ia.py"
+    start_dhcpd
+else
+    UHCPD="$(readlink -f "${USB_CDC_ECM_DIR}/../uhcpd/bin")/uhcpd"
+    start_uhcpd
+fi
```
Comment (benpicco) [inline dist/tools/usb-cdc-ecm/start_network.sh:68]: As soon as this part is reached, the `cleanup` function will be called.
When I ran 

```
% make BOARD=nrf52840dongle UPLINK=usb-cdc-ecm  term
make -C ethos
make -C uhcpd
make -C sliptty
make[2]: Nothing to be done for 'all'.
make[2]: Nothing to be done for 'all'.
make[2]: Nothing to be done for 'all'.
sudo sh /home/benpicco/dev/RIOT/dist/tools/usb-cdc-ecm/start_network.sh  2001:db8::/64
net.ipv6.conf.enx66410e90125b.forwarding = 1
net.ipv6.conf.enx66410e90125b.accept_ra = 0
Cleaning up...
```

was all I get - and 

```
2020-07-15 16:02:31,407 # uhcp_client(): sending REQ...
2020-07-15 16:02:41,406 # uhcp_client(): no reply received
2020-07-15 16:02:41,407 # uhcp_client(): sending REQ...
2020-07-15 16:02:51,406 # uhcp_client(): no reply received
```

on the device.

When I added a `sleep 10` here, it would work - but that's a hack.

```
2020-07-15 16:05:51,414 # got packet from fe80::1 port 48615
2020-07-15 16:05:51,415 # uhcp: push from fe80::1:48615 prefix=2001:db8::/64
2020-07-15 16:05:51,417 # gnrc_uhcpc: uhcp_handle_prefix(): add compression context 0 for prefix 2001:db8::ac0f:e90:125b:e86d/64
2020-07-15 16:05:51,420 # gnrc_uhcpc: uhcp_handle_prefix(): configured new prefix 2001:db8::ac0f:e90:125b:e86d/64
```

Also this claims to be `TERMPROG`, but it doesn't launch a terminal.
So why not kill two birds with one stone and launch `pyterm` here?

Still a bit of a hack, but that would get the desired result.

File: dist/tools/usb-cdc-ecm/start_network.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
+#!/bin/sh
+
+USB_CDC_ECM_DIR="$(dirname $(readlink -f $0))"
+
+INTERFACE="enp0s20f0u1"
```
[the remark this answers is not in the document]
Comment (benpicco) [reply dist/tools/usb-cdc-ecm/start_network.sh:5]: You could also try to auto-detect the interface name based on the contents of `/sys/bus/usb/drivers/cdc_ether/*/net/`

File: examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# USB Modules
+USEMODULE += usbus
```
[the remark this answers is not in the document]
Comment (trickeydan) [reply examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf]: d82d08f88

File: dist/tools/usb-cdc-ecm/start_network.sh (hunk 3 above)
Comment (trickeydan) [reply dist/tools/usb-cdc-ecm/start_network.sh:74]: 859eba150

Added an optional argument to `start_network.sh` so that a serial port can be specified and opened using `pyterm`. This is passed from `$PORT` in the Makefile, so that it always opens for the example. I specifically want this to be optional for "Headless" deployments where only the ethernet interface is connected, so we can just to `start_network $PREFIX` for those.

Hadn't realised that I'd removed the wait for key from this before submitting, as the `EXIT` does trigger the cleanup `trap`.

TLDR; A bit of a hack, but not the end of the world.

File: dist/tools/usb-cdc-ecm/start_network.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (benpicco) [reply dist/tools/usb-cdc-ecm/start_network.sh:5]: How about

```bash
INTERFACE=$(ls /sys/bus/usb/drivers/cdc_ether/*/net/ | cut -f1)
INTERFACE=${INTERFACE%/}
```

File: examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+USEMODULE += usbus_cdc_ecm
+
+ifeq (1,$(USE_DHCPV6))
+  FLAGS_EXTRAS += --use-dhcpv6
+endif
+
+# Configure terminal parameters for UHCP
+  TERMDEPS += host-tools  
+  TERMPROG ?= sudo sh $(RIOTTOOLS)/usb-cdc-ecm/start_network.sh
+  TERMFLAGS ?= $(FLAGS_EXTRAS) $(IPV6_PREFIX) $(PORT)
```
Comment (benpicco) [inline examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf]:
```suggestion
TERMDEPS += host-tools  
TERMPROG ?= sudo sh $(RIOTTOOLS)/usb-cdc-ecm/start_network.sh
TERMFLAGS ?= $(FLAGS_EXTRAS) $(IPV6_PREFIX) $(PORT)
```

no need to indent those

File: dist/tools/usb-cdc-ecm/start_network.sh [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+[ -z "${PREFIX}" ] && {
+    echo "usage: $0 [-d|--use-dhcpv6] <prefix> [<serial-port>]"
+    exit 1
+}
+
+if [ ! -z "$2" ]; then
+    PORT=$2
+fi
+
+
```
Comment (benpicco) [inline dist/tools/usb-cdc-ecm/start_network.sh:62]:
```suggestion
```
excessive whitespace

File: examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf (hunk 1 above)
Comment (trickeydan) [reply examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf]: 859eba1

File: examples/gnrc_border_router/Makefile (hunk 2 above)
Comment (trickeydan) [reply examples/gnrc_border_router/Makefile]: 12915f4ca

**Comment by trickeydan** [comment]
> but somehow uhcpc died along the way

I'm not seeing this issue. Do you have any info on how to replicate? I have a `nrf52840dongle` if that might be the issue?

File: examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf (hunk 6 above)
Comment (trickeydan) [reply examples/gnrc_border_router/Makefile.usb-cdc-ecm.conf]: 73d295f9f

File: dist/tools/usb-cdc-ecm/start_network.sh (hunk 7 above)
Comment (trickeydan) [reply dist/tools/usb-cdc-ecm/start_network.sh]: 73d295f9f

**Comment by benpicco** [comment]
Ah nvm, that was NetworkManager messing with the link.

File: dist/tools/usb-cdc-ecm/start_network.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (trickeydan) [reply dist/tools/usb-cdc-ecm/start_network.sh]: 09d7ed81c

**Comment by benpicco** [comment]
Please squash!
You want to chance to commit message to something like `examples/gnrc_border_router: add CDC-ECM uplink `

**Comment by benpicco** [comment]
What also (kinda) works is bridging the usb and the ethernet interface 

```
sudo ip link add name br0 type bridge
sudo ip link set br0 up
sudo ip link set eno1 master br0
sudo ip link set enx66410e90125b master br0 
```

When compiled with `USE_DHCPV6=1` (and not running a DHCPv6 server locally) the node will get a prefix from my local router and advertises it in the network.

```
2020-07-22 18:21:28,063 # Iface  7  HWaddr: E8:6D  Channel: 26  NID: 0x23 
2020-07-22 18:21:28,065 #           Long HWaddr: AE:0F:0E:90:12:5B:E8:6D 
2020-07-22 18:21:28,066 #            TX-Power: 0dBm L2-PDU:102  MTU:1280  HL:64  RTR  
2020-07-22 18:21:28,067 #           RTR_ADV  6LO  IPHC  
2020-07-22 18:21:28,068 #           Source address length: 8
2020-07-22 18:21:28,069 #           Link type: wireless
2020-07-22 18:21:28,071 #           inet6 addr: fe80::ac0f:e90:125b:e86d  scope: link  VAL
2020-07-22 18:21:28,072 #           inet6 addr: 2001:16b8:4534:cf8:ac0f:e90:125b:e86d  scope: global  VAL
2020-07-22 18:21:28,073 #           inet6 group: ff02::2
2020-07-22 18:21:28,073 #           inet6 group: ff02::1
2020-07-22 18:21:28,074 #           inet6 group: ff02::1:ff5b:e86d
2020-07-22 18:21:28,074 #           
2020-07-22 18:21:28,074 # Iface  6  HWaddr: 66:41:0E:90:12:5A 
2020-07-22 18:21:28,075 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2020-07-22 18:21:28,075 #           Source address length: 6
2020-07-22 18:21:28,076 #           Link type: wired
2020-07-22 18:21:28,076 #           inet6 addr: fe80::6441:eff:fe90:125a  scope: link  VAL
2020-07-22 18:21:28,077 #           inet6 group: ff02::2
2020-07-22 18:21:28,077 #           inet6 group: ff02::1
2020-07-22 18:21:28,077 #           inet6 group: ff02::1:ff90:125a
```

I can reach all devices in my network, but routing to the internet doesn't yet work with this.

**Comment by kmartinez** [comment]
Dan's thing is a nice move towards this mode of having nodes behind a firewall sending data to a server (for me that's CoAP of course!). I was just lucky to have a network-savvy student to work on it ;-)
